File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/controller/overlay/handler/CircleOverlayHandler.kt
package dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.handler

import com.naver.maps.map.overlay.CircleOverlay
import com.naver.maps.map.overlay.Overlay
import dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.OverlayHandler
import dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.OverlayHandler.Companion.getterName
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay.NCircleOverlay
import io.flutter.plugin.common.MethodChannel

internal interface CircleOverlayHandler : OverlayHandler {
    fun handleCircleOverlay(
        circleOverlay: Overlay, method: String, arg: Any?, result: MethodChannel.Result,
    ) = (circleOverlay as CircleOverlay).let { c ->
        when (method) {
            NCircleOverlay.centerName -> setCenter(c, arg!!)
            NCircleOverlay.radiusName -> setRadius(c, arg!!)
            NCircleOverlay.colorName -> setColor(c, arg!!)
            NCircleOverlay.outlineColorName -> setOutlineColor(c, arg!!)
            NCircleOverlay.outlineWidthName -> setOutlineWidth(c, arg!!)
            getterName(NCircleOverlay.boundsName) -> getBounds(c, result::success)
            else -> result.notImplemented()
        }
    }

    fun setCenter(circleOverlay: CircleOverlay, rawCenter: Any)

    fun setRadius(circleOverlay: CircleOverlay, rawRadius: Any)

    fun setColor(circleOverlay: CircleOverlay, rawColor: Any)

    fun setOutlineColor(circleOverlay: CircleOverlay, rawOutlineColor: Any)

    fun setOutlineWidth(circleOverlay: CircleOverlay, rawOutlineWidth: Any)

    fun getBounds(circleOverlay: CircleOverlay, result: (bounds: Map<String, Any>) -> Unit)
}

File: ios/flutter_naver_map/Sources/flutter_naver_map/controller/overlay/handler/CircleOverlayHandler.swift
import NMapsMap
import Flutter

internal protocol CircleOverlayHandler: OverlayHandler {
    func setCenter(_ circleOverlay: NMFCircleOverlay, rawCenter: Any)

    func setRadius(_ circleOverlay: NMFCircleOverlay, rawRadius: Any)

    func setColor(_ circleOverlay: NMFCircleOverlay, rawColor: Any)

    func setOutlineColor(_ circleOverlay: NMFCircleOverlay, rawOutlineColor: Any)

    func setOutlineWidth(_ circleOverlay: NMFCircleOverlay, rawOutlineWidth: Any)

    func getBounds(_ circleOverlay: NMFCircleOverlay, result: (_ bounds: Dictionary<String, Any>) -> Void)
}

internal extension  CircleOverlayHandler {
    func handleCircleOverlay(circleOverlay: NMFOverlay,
                             method: String,
                             args: Any?,
                             result: @escaping FlutterResult) {
        let circleOverlay = circleOverlay as! NMFCircleOverlay
        switch method {
        case NCircleOverlay.centerName: setCenter(circleOverlay, rawCenter: args!)
        case NCircleOverlay.radiusName: setRadius(circleOverlay, rawRadius: args!)
        case NCircleOverlay.colorName: setColor(circleOverlay, rawColor: args!)
        case NCircleOverlay.outlineColorName: setOutlineColor(circleOverlay, rawOutlineColor: args!)
        case NCircleOverlay.outlineWidthName: setOutlineWidth(circleOverlay, rawOutlineWidth: args!)
        case getterName(NCircleOverlay.boundsName): getBounds(circleOverlay, result: result)
        default: result(FlutterMethodNotImplemented)
        }
    }
}